File: src/lib/features/confidence.ts
// Confidence score calculation

export interface ConfidenceResult {
  confidence_score: number;
  missing_sources: string[];
  data_completeness: number;
}

export interface ConfidenceWeights {
  binance_spot: number;
  binance_futures: number;
  coingecko: number;
}

const DEFAULT_WEIGHTS: ConfidenceWeights = {
  binance_spot: 0.3,
  binance_futures: 0.4,
  coingecko: 0.3,
};

/**
 * Calculate confidence score based on data source availability
 */
export function calculateConfidence(
  binanceSpotOk: boolean,
  binanceFuturesOk: boolean,
  coingeckoOk: boolean,
  hasFutures: boolean,
  weights: ConfidenceWeights = DEFAULT_WEIGHTS
): ConfidenceResult {
  const missing: string[] = [];
  let spotW: number;
  let futW: number;
  let cgW: number;

  if (!hasFutures) {
    // Redistribute futures weight to other sources
    const totalW = weights.binance_spot + weights.coingecko;
    spotW = weights.binance_spot / totalW;
    cgW = weights.coingecko / totalW;
    futW = 0;
  } else {
    spotW = weights.binance_spot;
    futW = weights.binance_futures;
    cgW = weights.coingecko;
  }

  let score = 0;

  if (binanceSpotOk) {
    score += spotW * 100;
  } else {
    missing.push("binance_spot");
  }

  if (hasFutures) {
    if (binanceFuturesOk) {
      score += futW * 100;
    } else {
      missing.push("binance_futures");
    }
  }

  if (coingeckoOk) {
    score += cgW * 100;
  } else {
    missing.push("coingecko");
  }

  const totalSources = hasFutures ? 3 : 2;
  const dataCompleteness = ((totalSources - missing.length) / totalSources) * 100;

  return {
    confidence_score: Number(score.toFixed(1)),
    missing_sources: missing,
    data_completeness: Number(dataCompleteness.toFixed(1)),
  };
}

File: backend/features/confidence.py
"""
Confidence score calculation based on data source availability
"""
from typing import Dict, List


DEFAULT_WEIGHTS = {
    "binance_spot": 0.30,
    "binance_futures": 0.40,
    "coingecko": 0.30,
}


def calculate_confidence(
    binance_spot_ok: bool,
    binance_futures_ok: bool,
    coingecko_ok: bool,
    has_futures: bool,
    weights: Dict[str, float] = None,
) -> Dict[str, any]:
    """
    Calculate confidence score based on data source availability
    
    weights: { "binance_spot": 0.30, "binance_futures": 0.40, "coingecko": 0.30 }
    has_futures: False nếu coin không có perpetual → không phạt binance_futures
    """
    if weights is None:
        weights = DEFAULT_WEIGHTS.copy()

    missing: List[str] = []

    if not has_futures:
        # Redistribute futures weight to other sources
        total_w = weights["binance_spot"] + weights["coingecko"]
        spot_w = weights["binance_spot"] / total_w
        cg_w = weights["coingecko"] / total_w
        fut_w = 0.0
    else:
        spot_w = weights["binance_spot"]
        fut_w = weights["binance_futures"]
        cg_w = weights["coingecko"]

    score = 0.0

    if binance_spot_ok:
        score += spot_w * 100
    else:
        missing.append("binance_spot")

    if has_futures:
        if binance_futures_ok:
            score += fut_w * 100
        else:
            missing.append("binance_futures")

    if coingecko_ok:
        score += cg_w * 100
    else:
        missing.append("coingecko")

    total_sources = 3 if has_futures else 2
    data_completeness = ((total_sources - len(missing)) / total_sources) * 100

    return {
        "confidence_score": round(score, 1),
        "missing_sources": missing,
        "data_completeness": round(data_completeness, 1),
    }
